### DISCUSSION TITLE
bekomme fallweise keine Daten über MQTT

### DISCUSSION THREAD

**Discussion by best57** [tbnobody/OpenDTU#578] [Q&A]
Hallo, 
zu Beginn einmal vielen Dank für dieses tolle Projekt. Ich verwende es schon einige Monate. Die Hauptfunktion bei mir ist die Realisierung einer Nulleinspeisung. Das funktioniert soweit auch prima. Allerdings: Alle 1-2 Wochen passiert es, dass keine Daten über MQTT zum ioBroker gelangen. Am Vortag gibt es noch keine Probleme, am nächsten Tag dann keine Daten. Bei den MQTT Information ist dann der Connection Status auf disconnected. Bei den "Live Data" wird alles korrekt angezeigt. Wenn ich die MQTT-Settings öffne und den Save-Button drücke, gehts wieder bis zum nächsten Ausfall. Ich verwende aktuell die version 1d3acb4, habe aber dieses Verhalten schon bei den vorhergehenden Versionen beobachtet.
Gibt es dazu einen Hinweis, dem ich nachgehen kann?

**Comment by turbolift** [comment]
Ich hatte dies jetzt auch schon zweimal ... den Tipp mit dem Speichern werde ich beim nächsten Mal auch versuchen ... Version 422d489

**Comment by Hans-JR** [comment]
Auch bei mir ist das genauso passiert (Version 422d489). Habe dann auch nur den Save-Button gedrückt und es lief wieder.
Anbei meine Config (wenn alles OK ist).
![MQTT-Config](https://user-images.githubusercontent.com/37930673/218133952-540fb2f8-8a04-4d10-850a-4fb97ee83472.jpg)

**Comment by best57** [comment]
Nach meinem Posting, am Freitag, habe ich ein Update auf die aktuelle version b8ced54 gemacht. Seither ist jeden Morgen, der Ausfall passiert. Ein Klick auf den Save-Button und es läuft wieder.

**Comment by best57** [comment]
Ich habe nun dieses Problem vorerst gelöst, indem ich per '/api/mqtt/status' den Status hole und, falls 'disconnected' per '/api/mqtt/config' die Config neu gespeichert wird.

**Reply by Hans-JR** [reply]
Gute Idee. Würdest Du Dein Script hier teilen. Ich bin noch nicht so fit in Blockly oder JS.

**Comment by best57** [comment]
hier das Blockly in JS exportiert. 
IP und das password entsprechend anpassen:

```
var mqtt_status_pruefen, status_dtu, result;

function subsequenceFromStartLast(sequence, at1) {
  var start = at1;
  var end = sequence.length - 1 + 1;
  return sequence.slice(start, end);
}


mqtt_status_pruefen = schedule('* * * * *', async function () {
  // MQTT-Status abfragen
  try {
    require("request")('http://192.168.xx.xx/api/mqtt/status', async function (error, response, result) {
      status_dtu = subsequenceFromStartLast(result.split(',')[5], 17);
    console.log((['+++ MQTT-Status: ',status_dtu,' +++'].join('')));
    if (status_dtu == 'false') {
      // wenn connected = false
      console.log('+++ MQTT-Connection Status: disconnected +++');
      // config holen
      try {
        require("request")('http://admin:password@192.168.xx.xx/api/mqtt/config', async function (error, response, result) {
          // config speichern
        exec((['/usr/bin/curl -u "admin:password" http://192.168.xx.xx/api/mqtt/config -d data=\'',result,'\''].join('')), async function (error, result, stderr) {
            console.log((['+++ Result: ',result,' +++'].join('')));
        });
        }).on("error", function (e) {console.error(e);});
      } catch (e) { console.error(e); }
    }
    }).on("error", function (e) {console.error(e);});
  } catch (e) { console.error(e); }
});
```

**Comment by best57** [comment]
mittlerweile bin ich mir sicher, dieses Prolem tritt nur dann auf, wenn die WLAN-Verbindung instabil ist. Mit einer externen WLAN-Antenne ist nun die Verbindung stabiler und MQTT bleibt dauerhaft "connected".
